llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp

```diff
@@ -1087,24 +1135,44 @@ bool LoopIdiomRecognize::processLoopStridedStore(
         BasePtr, SplatValue, NumBytes, MaybeAlign(StoreAlignment),
         /*isVolatile=*/false, AATags.TBAA, AATags.Scope, AATags.NoAlias);
   } else {
-    assert (isLibFuncEmittable(M, TLI, LibFunc_memset_pattern16));
-    // Everything is emitted in default address space
-    Type *Int8PtrTy = DestInt8PtrTy;
-
-    StringRef FuncName = "memset_pattern16";
-    FunctionCallee MSP = getOrInsertLibFunc(M, *TLI, LibFunc_memset_pattern16,
-                            Builder.getVoidTy(), Int8PtrTy, Int8PtrTy, IntIdxTy);
-    inferNonMandatoryLibFuncAttrs(M, FuncName, *TLI);
-
-    // Otherwise we should form a memset_pattern16.  PatternValue is known to be
-    // an constant array of 16-bytes.  Plop the value into a mergable global.
-    GlobalVariable *GV = new GlobalVariable(*M, PatternValue->getType(), true,
-                                            GlobalValue::PrivateLinkage,
-                                            PatternValue, ".memset_pattern");
-    GV->setUnnamedAddr(GlobalValue::UnnamedAddr::Global); // Ok to merge these.
-    GV->setAlignment(Align(16));
-    Value *PatternPtr = GV;
-    NewCall = Builder.CreateCall(MSP, {BasePtr, PatternPtr, NumBytes});
+    assert(isLibFuncEmittable(M, TLI, LibFunc_memset_pattern16) ||
+           EnableMemsetPatternIntrinsic);
+    if (EnableMemsetPatternIntrinsic) {
+      // Everything is emitted in default address space
+
+      // Get or insert the intrinsic declaration
+      Function *MemsetPatternIntrinsic = Intrinsic::getDeclaration(
+          M, Intrinsic::memset_pattern,
+          {DestInt8PtrTy, Builder.getInt128Ty(), Builder.getInt64Ty()});
+
+      // Create the call to the intrinsic
+      NewCall = Builder.CreateCall(
+          MemsetPatternIntrinsic,
+          {BasePtr,
+           memSetPatternValueToI128ConstantInt(M->getContext(), PatternValue),
```

Comment:
Two high level suggestions to simplify this code:
* Change getMemSetPatternValue to directly return the i128 result.  There's no reason it should return and array, and then be converted again.  The global variable can just be a i128 type, and the pointer can be passed.  The Array is (should be?) an irrelevant implementation detail.
* Move the libcall creation to lowering for the intrinsic, and always generate the intrinsic.  The enable becomes not whether to use the intrinsic, but *when* to generate the intrinsic (i.e. are we limiting ourselves to the case where we know we have a libcall to expand to.)